design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,87 @@
+
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There should be option in `operator.yaml` to validate whether the COSI controller exists and another option at object store level to bring up the Ceph COSI driver. Each Ceph object store should have its own COSI driver.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+- Create Service Account and RBAC for COSI driver
+
+```bash
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml
+kubectl create -f https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml
+```
+
+### Changes in Operator.yaml
+
+Below option validates whether the COSI controller is deployed in the cluster. If it is not deployed, it will log an warning message and when user tries to create object store woth cosi, it will reconcile until the controller is deployed.
+
+```
+  cosiController:
+    enabled: true
+```
+
+### Changes in ObjectStore.yaml
+
+Following option will bring up the Ceph COSI driver for the RGW server. It will also create `CephObjectStoreUser`  objectstorage-provisioner which provides credentials for Ceph COSI driver.
+```
+  cephCosiDrvierSpec:
+    enabled: true
+```
+
+### How Rook can improve COSI driver reliability
+
+Rook can ensure that resources need for Ceph COSI driver such as Ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster which already present with current codebase. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for multiple ceph object stores in the cluster. Rook also need to support update of COSI-Controller and COSI-Driver images independently
+
+### Coexistence of COSI and libbucket provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.
```

Comment:
Why would we want to support the same backend bucket to be managed from COSI and OBC? Will that help us transition to COSI? If so, how about a new section that talks about transitioning from OBC to COSI?